raiden/raiden_service.py

```diff
@@ -619,6 +633,141 @@ def settle(self, token_address, partner_address):
         return netting_channel
 
 
+class ConnectionManager(object):
+    """
+    The ConnectionManager provides a high level abstraction for connecting to a
+    Token network.
+    """
+    I_AM_HERE_ADDR = '2' * 40
+
+    def __init__(
+        self,
+        raiden,
+        token_address,
+        initial_channel_target=3,  # number of channels to open immediately
+        joinable_funds_target=.4,  # amount of funds not initially assigned
+    ):
+        self.lock = Semaphore()
+        self.funds = 0
+        self.raiden = raiden
+        self.token_address = safe_address_decode(token_address)
+        self.initial_channel_target = initial_channel_target
+        self.joinable_funds_target = joinable_funds_target
+        # force state update
+        self.raiden.event_handler.poll_all_event_listeners()
+        try:
+            self.manager = self.raiden.get_manager_by_token_address(self.token_address)
+        except UnknownTokenAddress:
+            log.debug("token not yet registered", token=pex(self.token_address))
+            self.raiden.chain.default_registry.add_token(self.token_address)
+            self.raiden.event_handler.poll_all_event_listeners()
+            self.manager = self.raiden.get_manager_by_token_address(self.token_address)
+            # make ourselves visible
+            self.raiden.api.open(
+                self.token_address.encode('hex'),
+                ConnectionManager.I_AM_HERE_ADDR
+            )
+
+    def find_new_partners(self, number):
+        available = self.manager.channelgraph.graph.nodes()
+        if self.__class__.I_AM_HERE_ADDR.decode('hex') in available:
+            available.remove(self.__class__.I_AM_HERE_ADDR.decode('hex'))
+        if self.raiden.address in available:
+            available.remove(self.raiden.address)
+        for connected in self.open_channels:
+            available.remove(connected.partner_address)
+        available = self._select_best_partners(available)
+        return available[:number]
+
+    def _select_best_partners(self, partners):
+        # FIXME: use a proper selection strategy
+        return partners
+
+    def _initial_funding_per_partner(self):
+        return int(
+                self.funds * (1 - self.joinable_funds_target)
+                / self.initial_channel_target
+                )
+
+    def connect(self, funds):
+        """ Connect to the network """
+        with self.lock:
+            self.funds = funds
```

Comment:
This is confusing, `_initial_funding_per_partner` is relying on this state change. Having funds as a argument to `_initial_funding_per_partner` is more understandable.